lib/vaultex/client.ex

```diff
@@ -64,26 +61,107 @@ defmodule Vaultex.Client do
 
   """
   def read(key, auth_method, credentials) do
-    response = read(key)
-    case response do
-      {:ok, _} -> response
+    retry_with_auth auth_method, credentials, fn ->
+      GenServer.call(:vaultex, {:read, key})
+    end
+  end
+
+  @doc """
+  Writes a secret to Vault given a path.
+
+  ## Parameters
+
+  - key: A String path where the secret will be written.
+  - value: A String => String map that will be stored in Vault
+  - auth_method and credentials: See Vaultex.Client.auth
+
+  ## Examples
+
+  ```
+  iex> Vaultex.Client.write "secret/foo", %{"value" => "bar"}, :app_id, {app_id, user_id}
+  :ok
+  ```
+
+  """
+  def write(key, value, auth_method, credentials) do
+    retry_with_auth auth_method, credentials, fn ->
+      GenServer.call(:vaultex, {:write, key, value})
+    end
+  end
+
+  defp retry_with_auth(auth_method, credentials, fun) do
+    case fun.() do
       {:error, _} ->
         with {:ok, _} <- auth(auth_method, credentials),
-          do: read(key)
+          do: fun.()
+      response -> response
     end
   end
 
-  defp read(key) do
-    GenServer.call(:vaultex, {:read, key})
-  end
+  # GenServer callbacks
 
-  def handle_call({:read, key}, _from, state) do
-    Read.handle(key, state)
+  def init(state) do
+    {:ok, Map.merge(state, %{url: url()})}
   end
 
   def handle_call({:auth, method, credentials}, _from, state) do
-    Auth.handle(method, credentials, state)
+    {auth_path, credentials} = build_auth_params(method, credentials)
```

Comment:
Why is this move into here? Any reason why Auth can't stay in it's own abstracted module?